### PR TITLE
Add -Raw switch to Select-String for convenience

### LINKED ISSUE #7713: Add a switch to Select-String that makes it return just strings for convenience and performance
<sup>Related: #7712</sup>

Sometimes, all you're interested in is the matching input lines as _strings_ rather than full-blown `[Microsoft.PowerShell.Commands.MatchInfo]` instances.

Not having to extract the `.Line` property in subsequent processing can also significantly improve performance.

<sup>_Update_: Performance was originally discussed only with respect to how expensive wrapping the matched strings in `MatchInfo` instances is to begin with: nothing to worry about, apparently - see comments for a discussion.</sup>

A new switch named, say, `-Bare` switch could instruct `Select-String` to output (undecorated) _strings_ only.

```powershell
$res = 'line1', 'lineother', 'line3' | Select-String '\d' -Bare; $res; '---'; $res[0].GetType().Name
line1
line3
---
String
```

Note: I'm suggesting the somewhat abstract name `-Bare`, because the abstract logic of this proposal - namely to output "bare" objects that are _undecorated_ (have no ETS properties added to them) / are _not wrapped in instances of a helper type_ - applies to other cmdlets as well, such as in #7537, and its conceivable that other cmdlets may benefit from `-Bare` too, such as `ConvertTo-Json` in order to solve #5797; other cmdlets could benefit from such a switch too, such as `Compare-Object`.  
_Update_: The case for `-Bare` as a general pattern has since been made in #7855.

Environment data
----------------

Written as of:

```powershell
PowerShell Core 6.1.0-preview.4
```

### PR DESCRIPTION
**Pull request by Jawz84**
# PR Summary

Resolves #7713 
This adds a `-Raw` switch parameter to Select-String, for convenience. 

## PR Context

This is my first PR to the PowerShell repo, so I picked this tiny issue. As to why this switch is desirable has already been discussed thoroughly in the issue. 

~~The approach I've taken does not introduce new parameter sets, even though -Raw is not compatible with -Quiet and -Context.~~ In the case -Raw is used together with -Context, the latter is ignored. To me, it would not make sense to have context lines when all we want is matched lines. ~~In the case -Raw is used together with -Quiet, Quiet takes precedence, because the choice for precedence seemed somewhat arbitrary and it was easier to implement this way. One might argue that respecting a users whish for -Quiet behaviour is more important than returning strings in that case.~~
_Edit: after discussion, -Raw and -Quiet were put in separate parametersets._

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.

I'm unsure about this part below ⬇️ . I think you could say this is a backwards compatible breaking change. It introduces a new feature, but does not break existing functionality. Please advise.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [ ] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [x] Issue filed: https://github.com/MicrosoftDocs/PowerShell-Docs/issues/4537
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)

### CODE REVIEW

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/MatchString.cs
Comment (vexx32) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/MatchString.cs]: This should probably be in a separate parameter set to the `Quiet` switch, otherwise behaviour when both are supplied will be confusing regardless of which is provided.

They both transform the output, so we shouldn't allow them to be present at the same time imo. This would also let you simplify your logic further down a bit by checking `ParameterSetName` before looking at Raw/Quiet

**Comment by Jawz84** [comment]
I'll need some guidance as to what I am supposed to do to fix any failing CI requirements. Please let me know what I can do as a next step. Thank you 🙂.

**Review by vexx32 — COMMENTED** [review]
Love it! One thing to note but other than that this looks good. :smile:

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/MatchString.cs
Comment (Jawz84) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/MatchString.cs]: Thanks for spotting that, I've thought about this. Did you see my comment here: 

> The approach I've taken does not introduce new parameter sets, even though -Raw is not compatible with -Quiet and -Context. In the case -Raw is used together with -Context, the latter is ignored. To me, it would not make sense to have context lines when all we want is matched lines. In the case -Raw is used together with -Quiet, Quiet takes precedence, because the choice for precedence seemed somewhat arbitrary and it was easier to implement this way. One might argue that respecting a users whish for -Quiet behaviour is more important than returning strings in that case.

The -List and -Quiet params did not go in separate parameter sets either. That also made me swing this way.

Comment (vexx32) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/MatchString.cs]: I hear you, and maybe the PS team will have a different opinion, but in my mind it's a lot better from a UX standpoint to have these separated out in different sets. 

It's too easy to get confused as to which output takes precedence. We have the ability to separate them out, so I'd say... why not? 😄

Comment (Jawz84) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/MatchString.cs]: OK, no worries. I'll put -Raw and -Quiet in different paramsets then 😉 . First thing tomorrow morning!

Comment (SteveL-MSFT) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/MatchString.cs]: I would just say that don't assume the current code is the best way to do something :)  PowerShell is a large code base written by a large number of people over a large number of years.  It's generally rare that old code is updated when new and better patterns emerge.  In this specific case, my preference is to have different parameter sets and consider fixing `-List` and `-Quiet` (unless there's a clear reason why it was done this way, but doesn't have to be part of this PR, can be separate one).  Note that we do have specific cases where we decided it was best to not have parametersets and do a runtime check due to the complexity of many parameters that were mutually exclusive over several existing parametersets, but those should be exceptions.

Comment (vexx32) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/MatchString.cs]: @SteveL-MSFT I think `Where-Object` begs to differ. I've seen those very very naughty parameter sets 😉

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Select-String.Tests.ps1
Comment (SteveL-MSFT) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Select-String.Tests.ps1]: Why is there a `SLS` tag?  We're using specific tags that our CI understands: https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/WritingPesterTests.md#describecontextit

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Select-String.Tests.ps1
Comment (SteveL-MSFT) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Select-String.Tests.ps1]:
```suggestion
	It "Should return a string type when -Raw is used" {
```

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Select-String.Tests.ps1
Comment (SteveL-MSFT) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Select-String.Tests.ps1]: Why are you explicitly changing the left side to an array?

**Review by SteveL-MSFT — CHANGES_REQUESTED** [review]

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Select-String.Tests.ps1
Comment (Jawz84) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/Select-String.Tests.ps1]: I'll remove that - was there to shorten my test cycles. Has slipped in by accident.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/MatchString.cs
Comment (Jawz84) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/MatchString.cs]: param sets it is!

**Comment by Jawz84** [comment]
I'll have to finish reading rest of the comments later (I have an appointment tonight). Thank you for getting back to me so fast! Much appreciated!

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Select-String.Tests.ps1
Comment (Jawz84) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/Select-String.Tests.ps1]: Tbh: because the other tests that were already there were set up like that too, and I did not want to break style. I'll remove it. It does not matter for the outcome of the test though.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Select-String.Tests.ps1
Comment (Jawz84) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/Select-String.Tests.ps1]: fixed.

**Comment by Jawz84** [comment]
Parametersets and accompanying test fixed. Just let me know if there is anything else that needs to be done 😉

**Comment by daxian-dbw** [comment]
Close and reopen the PR to force Windows CI build working properly.

**Comment by Jawz84** [comment]
Hi Dongbo @daxian-dbw, I didn't know the trick with close/reopen issue to get CI to rerun. Would you like me to do that again? This time, the Mac OS test lost connection with the agent.

**Comment by Jawz84** [comment]
I'll close and reopen this once, to re-trigger CI, this clearly looks like a glitch. If that doesn't turn out all- green, I leave it to the team.

**Comment by daxian-dbw** [comment]
@PoshChan please get failures for macOS

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/MatchString.cs
Comment (PoshChan) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/MatchString.cs]: @daxian-dbw, test results were not published for PowerShell-CI-linux at vstfs:///Build/Build/25352

Comment (PoshChan) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/MatchString.cs]: @daxian-dbw, your last commit had 1 failures in `PowerShell-CI-macos`
Read-Host.[Console]::ReadKey() implementation on non-Windows.Should output correctly
```powershell
Expected 0, but got 1.
at <ScriptBlock>, /Users/vsts/agent/2.153.2/work/1/s/test/powershell/Host/Read-Host.Tests.ps1: line 25
25:             $LASTEXITCODE | Should -Be 0
```

Comment (PoshChan) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/MatchString.cs]: @daxian-dbw, test results were not published for PowerShell-CI-static-analysis at vstfs:///Build/Build/25389

Comment (PoshChan) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/MatchString.cs]: @daxian-dbw, test results were not published for PowerShell-CI-windows at vstfs:///Build/Build/25392

Comment (daxian-dbw) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/MatchString.cs]: @PoshChan Please rerun for macOS

**Comment by Jawz84** [comment]
Mac OS failed again, where it passed the other day. Awaiting the PSTeam.

**Comment by Jawz84** [comment]
This PR seems blocked on CI glitches. Could you please have a look @anmenaga? If there is anything I can do, just let me know :). Thank you!

**Comment by Jawz84** [comment]
@PoshChan Please rerun for macOS

**Comment by PoshChan** [comment]
@Jawz84, you are not authorized to request a rebuild

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/MatchString.cs
Comment (Jawz84) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/MatchString.cs]: @SteveL-MSFT I've implemented the requested changes. CI is happy now. I'm not sure how to mark your requested change as 'done' :)

**Comment by Jawz84** [comment]
The requested changes have been implemented. CI is all green. @stevel_msft can you have another look? I think this is ready.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/MatchString.cs
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/MatchString.cs]: Please remove the obvious comment.
```suggestion
```

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Select-String.Tests.ps1
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Select-String.Tests.ps1]:
```suggestion
	    $equal | Should -BeTrue
```

**Comment by Jawz84** [comment]
Thank you for your review @iSazonov.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Select-String.Tests.ps1
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Select-String.Tests.ps1]: Please use BeExactly for strings. Below too.
```suggestion
	    (Select-String in $testInputFile -Raw)[0] | Should -BeExactly $expected1
```

Comment (Jawz84) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/Select-String.Tests.ps1]: I don't think that is the right thing to do in this case: Select-String is case insensitive, unless `-CaseSensitive` is specified. Are you sure?

Comment (iSazonov) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/Select-String.Tests.ps1]: We do not check "case insensitive". The check should guarantee that we received _exactly_ the original string. Perhaps it is not critical for the test but it is best practice.

Comment (Jawz84) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/Select-String.Tests.ps1]: OK. I see your point, and I agree now. Would you like me to retrofit this same fix to lines 133-137?

Comment (iSazonov) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/Select-String.Tests.ps1]: You can do this in the PR (or move to follow PR).

**Review by iSazonov — APPROVED** [review]

**Comment by iSazonov** [comment]
@SteveL-MSFT Please update your review.

**Review by SteveL-MSFT — APPROVED** [review]

**Comment by anmenaga** [comment]
@Jawz84 :
* Please open a [documentation issue](https://github.com/powershell/powershell-docs/) to document new parameter and updates to parameter sets; Please add a link to that doc issue to the description of this PR under "Issue filed:".
* The CodeFactor issue in the new code needs to be fixed - `Insert parentheses within the conditional AND and OR expressions to declare the operator precedence.` - line 1186 - `private IContextTracker GetContextTracker() => (Raw || _preContext == 0 && _postContext == 0) ? _noCon..`

@vexx32 Can you please look through this PR again and approve if you don't see anything blocking? Thank you.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/MatchString.cs
Comment (vexx32) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/MatchString.cs]: Why are these sets named `ObjectQuiet` and `FileQuiet`? Isn't the normal behaviour with these parameter sets to output the MatchInfo objects? 

It should only need to be named after the `-Quiet` switch if that switch is the defining mandatory parameter for that set -- in this case it's just an optional switch for any of these sets.

I'd recommend just using `Object` / `File` / `ObjectRaw` / `FileRaw` names, as the `Quiet` appears to imply the default behaviour is `-Quiet`, which would probably not be desirable for future maintainability. 🙂

**Review by vexx32 — COMMENTED** [review]
Apart from the parameter set names I think this is all good to go.

Nice work! 😄

**Comment by Jawz84** [comment]
@anmenaga Thank you, I've added parens to fix the CodeFactor issue. I've also put in a rudimentary issue in for documentation, and added it to the description above.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/MatchString.cs
Comment (Jawz84) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/MatchString.cs]: I'll have to look into this one later, thank you.

Comment (Jawz84) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/MatchString.cs]: Yes, absolutely. After a good night's sleep, this makes lots of sense now. I've removed the 'Quiet' part of the parameternames, as you suggested. Good catch.

**Comment by Jawz84** [comment]
Closing/reopening this to trigger CI. Mac os test ran into an unrelated error: 
```text
at <ScriptBlock>, /Users/vsts/agent/2.153.2/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Core/Job.Tests.ps1: line 108
108:                         $results = Receive-Job -Keep $job
```
https://powershell.visualstudio.com/PowerShell/_build/results?buildId=27521&view=ms.vss-test-web.build-test-results-tab&runId=1142426&resultId=104912&paneView=debug

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/MatchString.cs
Comment (vexx32) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/MatchString.cs]: If you look at `TestConnectionCommand` for a moment, it has a great example of how to work with OutputType -- you can actually specify the output type(s) based on the parameter set name.

I would redefine it to have the MatchInfo/bool type in Object & File sets, and then give it string output type for the ObjectRaw / FileRaw sets. :)

**Review by vexx32 — APPROVED** [review]
One minor nit, but otherwise this looks awesome. Great work! 😊

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/MatchString.cs
Comment (Jawz84) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/MatchString.cs]: Oh, I did not know that was even possible! Neat! Will look at that, thanks.

Comment (anmenaga) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/MatchString.cs]: So looks like this is the only remaining thing for this PR and then we're merging it.  or let me know if this is not blocking.

Comment (Jawz84) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/MatchString.cs]: I'm going to work on this tonight. I have no idea if this would be blocking or not. That's not up to me either I think 🙂

Comment (Jawz84) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/MatchString.cs]: @vexx32, I've pushed changes, could you have a look please? The problem I ran into is, that because `-Quiet` is not in it's own paramter set, I don't think we can get `OutputType` to be more specific than this. What do you think?

Comment (Jawz84) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/MatchString.cs]: I've also gone ahead and refactored the parameter set names to string constants, just like in TestConnectionCommand.cs. It makes the code more readable and easier to refactor later.

**Comment by Jawz84** [comment]
Close/Reopen issue to trigger CI. CI encountered a transient error, `Test-Connection` timeout:
`Expected the actual value to be less than 3, but got 3.0422223.`

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/MatchString.cs
Comment (vexx32) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/MatchString.cs]: `-Raw` and `-Quiet` can't be used together, right? I'd think this can just be `typeof(string)` as you shouldn't run into bool here.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/MatchString.cs
Comment (vexx32) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/MatchString.cs]: Since we're adding something here anyway, I feel it might be time to tidy this long line up a bit?

```suggestion
        private IContextTracker GetContextTracker() => (Raw || (_preContext == 0 && _postContext == 0)) ? _noContextTracker : new ContextTracker(_preContext, _postContext);
```
```suggestion
        private IContextTracker GetContextTracker() => (Raw || (_preContext == 0 && _postContext == 0)) 
            ? _noContextTracker
            : new ContextTracker(_preContext, _postContext);
```

**Review by vexx32 — COMMENTED** [review]
Definitely looking much better! 🙂

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/MatchString.cs
Comment (Jawz84) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/MatchString.cs]: Agreed, fixed.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/MatchString.cs
Comment (Jawz84) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/MatchString.cs]: Huh I left that in? Thanks, fixed.

**Comment by Jawz84** [comment]
@anmenaga over to you 🙂

**Comment by anmenaga** [comment]
@Jawz84 Thank you for your contribution!

**Pull request merged**
